File: Source.Core/Code/_Jacobi.Vst.Interop/Plugin/Configuration.cpp
#include "pch.h"
#include "Configuration.h"

#using <Microsoft.Extensions.Configuration.dll>
#using <Microsoft.Extensions.Configuration.Abstractions.dll>
#using <Microsoft.Extensions.Configuration.FileExtensions.dll>
#using <Microsoft.Extensions.Configuration.Json.dll>

namespace Jacobi {
namespace Vst {
namespace Interop {
namespace Plugin {

Configuration::Configuration(System::String^ filePath)
{
	_basePath = filePath;
}

void Configuration::EnsureConfig()
{
	if (_config == nullptr)
	{
		auto builder = gcnew Microsoft::Extensions::Configuration::ConfigurationBuilder();
		Microsoft::Extensions::Configuration::FileConfigurationExtensions::SetBasePath(builder, _basePath);
		Microsoft::Extensions::Configuration::JsonConfigurationExtensions::AddJsonFile(builder, "vstsettings.json", true);
		_config = builder->Build();
	}
}

System::String^ Configuration::GetAppSetting(System::String^ key)
{
	EnsureConfig();

	if (_config != nullptr)
	{
		return _config->default[key];
	}

	return nullptr;
}

}}}} // Jacobi::Vst::Interop::Plugin

File: Source.Core/Code/Jacobi.Vst.Core/Plugin/Configuration.cs
using Microsoft.Extensions.Configuration;
using System;

namespace Jacobi.Vst.Core.Plugin
{
    /// <summary>
    /// VST.NET specific plugin configuration options
    /// </summary>
    public sealed class Configuration
    {
        private readonly string _basePath;
        private IConfigurationRoot _config;

        /// <summary>
        /// Open configuration based on the plugin folder.
        /// </summary>
        /// <param name="basePath">Must not be null.</param>
        public Configuration(string basePath)
        {
            _basePath = basePath ?? throw new ArgumentNullException(nameof(basePath));
        }

        /// <summary>
        /// Access the plugin config. Returns null if no config is found.
        /// </summary>
        [CLSCompliant(false)]
        public IConfiguration PluginConfig
        {
            get { EnsureConfig(); return _config; }
        }

        private void EnsureConfig()
        {
            if (_config == null)
            {
                _config = new ConfigurationBuilder()
                    .SetBasePath(_basePath)
                    .AddJsonFile("vstsettings.json", true)
                    .Build();
            }
        }

        private string GetAppSetting(string key)
        {
            EnsureConfig();

            if (_config != null)
            {
                return _config[key];
            }

            return null;
        }

    }
}
